File: src/main/java/org/opendsb/routing/remote/RemotePeerConnection.java
package org.opendsb.routing.remote;

import java.util.concurrent.CompletableFuture;

public class RemotePeerConnection {
	
	private RemotePeer remotePeer;
	
	public RemotePeerConnection(RemotePeer remotePeer) {
		super();
		this.remotePeer = remotePeer;
	}
	
	public RemotePeer getRemotePeer() {
		return remotePeer;
	}

	public boolean isWireConnected() {
		return remotePeer.isWireConnected();
	}
	
	public boolean isBusConnected() {
		return remotePeer.isBusConnected();
	}

	public synchronized CompletableFuture<Void> whenConnected() {
		
		CompletableFuture<Void> connectedFuture = new CompletableFuture<>();
		
		if(remotePeer != null && remotePeer.isBusConnected()) {
			 connectedFuture.complete(null);
		} else {
			remotePeer.addConnectedFuture(connectedFuture);
		}
		
		return connectedFuture;
	}
	
	public CompletableFuture<Void> whenDisconnected() {
		
		CompletableFuture<Void> disconnectedFuture = new CompletableFuture<>();

		if(remotePeer == null || !remotePeer.isBusConnected()) {
			 disconnectedFuture.complete(null);
		} else {
			remotePeer.addDisconnectedFuture(disconnectedFuture);
		}
		
		return disconnectedFuture;
	}

}

File: python/opendsb/src/opendsb/routing/remote/remotepeerconnection.py
from concurrent.futures import Future

from opendsb.routing.remote.remotepeer import RemotePeer


class RemotePeerConnection:
    def __init__(self, remote_peer: RemotePeer):
        self.remote_peer = remote_peer

    def when_connected(self) -> Future:
        connected_future = Future()
        if self.remote_peer is not None and self.remote_peer.is_bus_connected():
            connected_future.set_result(None)
        else:
            self.remote_peer.add_connected_future(connected_future)
        return connected_future

    def when_disconnected(self) -> Future:
        disconnected_future = Future()
        if self.remote_peer is None or not self.remote_peer.is_bus_connected():
            disconnected_future.set_result(None)
        else:
            self.remote_peer.add_disconnected_future(disconnected_future)
        return disconnected_future
